### ISSUE TITLE
Karma-Jasmine Returning "Unhandled Promise rejection" on Task: "jasmine.onComplete" with async/forEach

### ISSUE THREAD

**Issue by maplion** [karma-runner/karma#3124]
### Expected behaviour
No errors returned.

### Actual behaviour
Code and logic work fine and I get the proper output in the html reporter, but my console is spammed with errors, such as:

```
ERROR: 'Unhandled Promise rejection:', 'Cannot read property 'forEach' of undefined', '; Zone:', 'ProxyZone', '; Task:', 'jasmine.onComplete', '; Value:', TypeError: Cannot read property 'forEach' of undefined
TypeError: Cannot read property 'forEach' of undefined
```

### Environment Details
Angular CLI: 6.1.1
Node: 8.9.4
OS: win32 x64
Angular: 6.1.1
... animations, cli, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.7.4
@angular-devkit/build-angular     0.7.4
@angular-devkit/build-optimizer   0.7.4
@angular-devkit/build-webpack     0.7.4
@angular-devkit/core              0.7.4
@angular-devkit/schematics        0.7.1
@ngtools/webpack                  6.1.4
@schematics/angular               0.7.1
@schematics/update                0.7.1
rxjs                              6.2.2
typescript                        2.7.2
webpack                           4.9.2

- Karma version (output of `karma --version`):
Karma version: 3.0.0

- Relevant part of your `karma.config.js` file

### Steps to reproduce the behaviour

```typescript
    it('checks if json keys are missing', () => {
        return new Promise(async (resolve, reject) => {
            await jsonFileList1.forEach(async (set1) => {
                await set1.forEach(async (fileLeft) => {
                    await set1.forEach(async (fileRight) => {
                        const diff = deepDiff.diff(fileLeft, fileRight);
                        await diff.forEach(async (element) => {
                            expect(element.kind).not.toMatch(
                                'N', 'path: '.concat(element.path.toString())
                            );
                        });
                    });
                });
            },
                resolve()
            );
        });
    });
```

Note again that all of this works as expected and the output in the HTML is as expected; but the console is spammed with errors seemingly caused by the jasmine.onComplete function not properly reading the value of the forEach, so the forEach throws the exception "Cannot read property 'forEach' of undefined', which then causes and 'Unhandled Promise rejection".

Is there something I am missing or is this an issue with the framework?  Thank you.

**Comment by johnjbarton** [comment]
I suggest you start over with your test function and avoid using `await arr.forEach()`, it does not make any sense to await on a synchronous function. Also avoid mixing `new Promise()` and  `await/async`; it is rarely useful. Shoot for `it('checks if json keys are missing', async () => {` instead. Check each layer to be sure you really need await/async.   Checkout `Promise.all()`.

Anyway this is not an issue karma-runner can fix.

**Comment by maplion** [comment]
Thank you for the response johnjbarton.  I went the async await route on the forEach because I have experienced multiple times in javascript where it seems the forEach is not truly synchronous.  For example, in this case, without it, the value of the array is undefined at the the time it is hit and the function fails -- with it, it works (not sure if this is because it relies on json imports and if those are asyncronous or what).  I mixed in the Promise because I was making every kind of construction I could find and the spyOn() function was the only thing that successfully made the errors go away in its current construction; but this broke the functionality of the testing (was always successful).  I was then attempting to see if I could wrap it in a Promise to make it happy, and it didn't.  After trying everything, I came here with my latest amalgam.

**Comment by maplion** [comment]
After researching further, while `forEach` is synchronous/blocking, there is an issue with asynchronous content within the `forEach` and it seems that in my case, my json imports act as asynchronous, which is why my code doesn't run unless I have the async/awaits.  Here is an article that accurately describes the issue I expereience:  https://codeburst.io/javascript-async-await-with-foreach-b6ba62bbf404

While debugging and breakpoints seem quite unreliable and buggy inside tests (at least in VSCode), after setting up an asynchronous `forEach` as described in the above example, I was able to get an intuitive sense for what jasmine seems to be doing:  it seems that Jasmine runs the code, but at some point does a scan to find the "expects" or something and on that separate run through [based on the error, has to do with jasmine.onComplete], the array is undefined and throws the related TypeError, which the Jasmine code is not properly handling, so it throws an `Unhandled Promise rejection`.

I was able to resolve this by wrapping my asyncForEach function in a null check:

```javascript
export async function asyncForEachForJasmine(array, callback) {
    if (array) {
        for (let index = 0; index < array.length; index++) {
            await callback(array[index], index, array);
        }
    }
}
```

This allows both my code to run correctly (as it did before), but removes the error spam due to the asynchronous-jasmine issue.
